src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/fs/TeeStep.java

```diff
@@ -89,50 +103,92 @@ public boolean start() throws Exception {
-                private static final long serialVersionUID = 1L;
-                @Override
-                public OutputStream invoke(File f, VirtualChannel channel) throws IOException, InterruptedException {
-                    f = f.getAbsoluteFile();
-                    if (!f.getParentFile().exists() && !f.getParentFile().mkdirs()) {
-                        throw new IOException("Failed to create directory " + f.getParentFile());
-                    }
-                    try {
-                        return new RemoteOutputStream(Files.newOutputStream(f.toPath(), StandardOpenOption.CREATE, StandardOpenOption.APPEND/*, StandardOpenOption.DSYNC*/));
-                    } catch (InvalidPathException e) {
-                        throw new IOException(e);
-                    }
+
+            @Override
+            public RemoteOutputStream invoke(File f, VirtualChannel channel) throws IOException, InterruptedException {
+                f = f.getAbsoluteFile();
+                if (!f.getParentFile().exists() && !f.getParentFile().mkdirs()) {
+                    throw new IOException("Failed to create directory " + f.getParentFile());
+                }
+                try {
+                    OutputStream os = Files.newOutputStream(f.toPath(),
+                            StandardOpenOption.CREATE,
+                            append ? StandardOpenOption.APPEND : StandardOpenOption.TRUNCATE_EXISTING,
+                            StandardOpenOption.WRITE);
+
+                    return new RemoteOutputStream(os);
+                } catch (InvalidPathException e) {
+                    throw new IOException(e);
                 }
-            });
+            }
+
+        }
+
+        private RemoteOutputStream getStream() throws IOException, InterruptedException {
+            if (stream == null) {
+                stream = f.act(new TeeFile(append));
+                // From now on, when resumed and the stream gets recreated append new data.
+                append = true;
+            }
+            return stream;
         }
+
+        private void writeObject(ObjectOutputStream oos) throws IOException, InterruptedException {
+            /**
+             * This gets called either to serialize to allow resume or to transfer to a node.
+             * When resuming, the stream gets recreated.
+             * When transferring the stream needs to transfer as well.
+             */
+            oos.defaultWriteObject();
+            if (Channel.current() == null) {
+                // Serialization for resume.
+                oos.writeBoolean(false);
+            } else {
+                // Serialization for transfer.
+                oos.writeBoolean(true);
+                oos.writeObject(getStream());
+            }
+        }
+
+        private void readObject(ObjectInputStream ois) throws IOException, ClassNotFoundException {
+            ois.defaultReadObject();
+            if (ois.readBoolean()) {
```

Comment:
This scheme has a minor problem: it will throw an error when a `tee` step is resumed across an upgrade of the plugin, since it changes the serial form even in `program.dat` where that must not be done.

The issue could be fixed, I think, by using `writeReplace` and `readResolve` rather than `writeObject` and `readObject`: when `Channel.current() == null`, just replace with `this`; when replacing for Remoting, use a special struct class which just resolves back to a filter with the stream configured.

Reply:
By the way these kinds of bugs can generally be reproduced, when anyone wants to bother, in a test using `@LocalData`. In the case of the `tee` step that would be hard since the serial form in `program.dat` encodes an absolute file path, which would be invalid when running the test later. Testing resumption of `sh` steps suffers from the same problem. `JenkinsRule` offers no good way around this, though someday I might write a special test-only `PickleFactory` which replaces any `File`, `Path`, or `String` mentioning a file path starting with `$JENKINS_HOME` or an agent `remoteFS` root.